src/main/kotlin/org/kamiblue/client/util/CircularArray.kt

```diff
@@ -0,0 +1,56 @@
+package org.kamiblue.client.util
+
+class CircularArray<T : Number> constructor(private val array: Array<T>, filled: Boolean) : Iterable<T> {
+    private var _size = if (filled) array.size else 0
+    private var tail = -1
+
+    companion object {
+        inline fun <reified T : Number> create(size: Int, fill: T) = CircularArray(Array(size) { fill }, true)
+        inline fun <reified T : Number> create(size: Int) = CircularArray(Array(size) { 0 as T }, false)
+    }
+
+    val head: Int
+        get() =
+            if (_size == array.size)
+                (tail + 1) % array.size
+            else
+                0
+
+    val size: Int get() = _size
+
+    fun add(item: T) {
+        tail = (tail + 1) % array.size
+        array[tail] = item
+        if (_size < array.size) _size++
+    }
+
+    fun reset() {
+        this._size = 0
+        this.tail = 0
+    }
+
+    fun average(): Float {
+        var sum = 0.0
+        for (i in 0 until size) {
+            sum += array[i].toDouble()
+        }
+        return if (_size == 0) 0.0f else (sum / size).toFloat()
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private operator fun get(index: Int): T =
+        when {
+            _size == 0 || index > _size || index < 0 -> throw IndexOutOfBoundsException("$index")
+            _size == array.size -> array[(head + index) % array.size]
+            else -> array[index]
+        }
+
+    private class CircularArrayIterator<T : Number>(private val self: CircularArray<T>) : Iterator<T> {
```

Comment:
This can be a inner class so you don't need the `self` field to store reference to the CircularArray

Reply:
I forgot about inner classes. I threw this in after my very rough attempt at generics. I'll fix it up and make it a bit more polished.